applications/nrf_desktop/src/hw_interface/selector_hw.c

```diff
@@ -0,0 +1,308 @@
+			 "Maximum number of selector states is 255");
+	__ASSERT_NO_MSG(cur_selector_state < ARRAY_SIZE(pins));
+
+	struct selector_event *event = new_selector_event();
+
+	event->src = SELECTOR_SRC_HW;
+	event->state = cur_selector_state;
+
+	EVENT_SUBMIT(event);
+}
+
+static int read_state(void)
+{
+	u32_t value;
+
+	for (size_t i = 0; i < ARRAY_SIZE(pins); i++) {
+		int err = gpio_pin_configure(gpio_dev[pins[i].port],
+					     pins[i].pin,
+					     GPIO_PUD_PULL_DOWN | GPIO_DIR_IN);
+		if (err) {
+			LOG_ERR("Cannot configure pin (err %d)", err);
+			return err;
+		}
+
+		err = gpio_pin_read(gpio_dev[pins[i].port], pins[i].pin,
+				    &value);
+
+		if (err) {
+			LOG_ERR("Cannot read value from port:%" PRIu8
+				" pin: %" PRIu8, pins[i].port, pins[i].pin);
+			return err;
+		}
+
+		if (value) {
+			cur_selector_state = i;
+			return 0;
+		}
+	}
+
+	LOG_ERR("Switch state is unknown");
+	return -1;
+}
+
+static void selector_isr(struct device *dev, struct gpio_callback *cb,
+			 u32_t pins_mask)
+{
+	u8_t port = ARRAY_SIZE(port_map);
+
+	for (size_t i = 0; i < ARRAY_SIZE(port_map); i++) {
+		if (dev == gpio_dev[i]) {
+			port = i;
+			break;
+		}
+	}
+
+	__ASSERT_NO_MSG(port < ARRAY_SIZE(port_map));
+
+	for (size_t i = 0; i < ARRAY_SIZE(pins); i++) {
+		int err = gpio_pin_disable_callback(gpio_dev[pins[i].port],
+						    pins[i].pin);
+
+		if (err) {
+			LOG_ERR("Cannot disable callback (err %d)",
+				err);
+		}
+
+		if ((pins[i].port == port) && (pins_mask & BIT(pins[i].pin))) {
+			/* Only one pin may be set at the time. */
+			__ASSERT_NO_MSG((pins_mask & ~BIT(pins[i].pin)) == 0);
+			cur_selector_state = i;
+		}
+	}
+
+	k_delayed_work_submit(&selector_work, 0);
+}
+
+static int setup_interrupts(void)
+{
+	int key = irq_lock();
+
+	for (size_t i = 0; i < ARRAY_SIZE(pins); i++) {
+		if (i == cur_selector_state) {
+			continue;
+		}
+
+		int err = gpio_pin_configure(gpio_dev[pins[i].port],
+					     pins[i].pin,
+					     GPIO_PUD_PULL_DOWN | GPIO_DIR_IN |
+					     GPIO_INT | GPIO_INT_LEVEL |
+					     GPIO_INT_ACTIVE_HIGH);
+
+		if (err) {
+			LOG_ERR("Cannot configure pin (err %d)", err);
+			return err;
+		}
+
+		err = gpio_pin_enable_callback(gpio_dev[pins[i].port],
+					       pins[i].pin);
+		if (err) {
+			LOG_ERR("Cannot enable callback (err %d)", err);
+			return err;
+		}
+	}
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+static int disable_interrupts(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(pins); i++) {
+		int err = gpio_pin_disable_callback(gpio_dev[pins[i].port],
+						    pins[i].pin);
+
+		if (err) {
+			LOG_ERR("Cannot disable callback (err %d)", err);
+			return err;
+		}
+
+		err = gpio_pin_configure(gpio_dev[pins[i].port], pins[i].pin,
+					 GPIO_DIR_IN | GPIO_PUD_PULL_DOWN);
+		if (err) {
+			LOG_ERR("Cannot configure pin (err %d)", err);
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+void selector_work_fn(struct k_work *w)
+{
+	selector_event_send();
+	int err = setup_interrupts();
```

Comment:
In selector_isr you have disabled isr callbacks only, but here you call setup for interrupts (as at the hardware first setup, event interrupts are enabled, just callbacks are disabled). I suggest to do enable callbacks here.

Reply:
In `setup_interrupts` I just configure pins and enable_callbacks for them. Still, function name may be misleading - I will rename it to `enable_interrupts`.

Reply:
I do not understand. I've pointed that setup_interrupts does to much - more than is required to be done here. If interrupts are enabled and just callbacs disabled then why configure interrupts again? Or have I understood something in a wrong way?

Reply:
Currrently it is only enabling interrupts here. Pins reconfiguration was removed (as requested in other comment).